doc/ref_arch/kubernetes/chapters/chapter04.md

```diff
@@ -79,8 +99,16 @@ In order for the Kubernetes components to be conformant with the Reference Archi
 |`ra2.k8s.004`|Highly available worker nodes|An implementation must consist of at least one worker node per availability zone or fault domain to ensure the high availability and resilience of workloads managed by Kubernetes|[req.gen.rsl.01 req.gen.avl.01 req.kcm.gen.02](./chapter02.md#23-kubernetes-architecture-requirements)|
 |`ra2.k8s.005`|Kubernetes API Version|In alignment with the [Kubernetes version support policy](https://kubernetes.io/docs/setup/release/version-skew-policy/#supported-versions), an implementation must use one of three latest minor versions (`n-2`). e.g. if the latest version is 1.17 then the RI must use either 1.17, 1.16 or 1.15.|TBC|
-|`ra2.k8s.006`||||
-|`ra2.k8s.007`||||
+|`ra2.k8s.006`|NUMA Support|The `TopologyManager` and `CPUManager` feature gates must be enabled and configured on the kubelet (note, TopologyManager is enabled by default in Kubernetes v1.18 and later, with CPUManager enabled by default in Kubernetes v1.10 and later). `--feature-gates="...,TopologyManager=true,CPUManager=true" --topology-manager-policy=single-numa-node --cpu-manager-policy=static`|[e.cap.007](./chapter02.md#221-cloud-infrastructure-software-profile-capabilities) [infra.com.cfg.002](./chapter02.md#223-cloud-infrastructure-software-profile-requirements) [infra.hw.cpu.cfg.004](./chapter02.md#224-cloud-infrastructure-hardware-profile-requirements)|
+|`ra2.k8s.007`|DevicePlugins Feature Gate|In order to support the various acceleration devices required, the DevicePlugins feature gate must be enabled (note, this is enabled by default in Kubernetes v1.10 or later). `--feature-gates="...,DevicePlugins=true,..."`|Various, e.g. [e.cap.013, e.cap.014](./chapter02.md#221-cloud-infrastructure-software-profile-capabilities)|
+|`ra2.k8s.008`|System Resource Reservations|To avoid resource starvation issues on nodes, reserve compute resources for system daemons and Kubernetes system daemons such as kubelet, container runtime, etc. (requires Kubernetes version 1.17 or later). Use the following kubelet flags: `--reserved-cpus=[0-3]`|NB: There doesn't appear to be a requirement that this is for?  Do we need to be specific about the values used?  The guidelines [here](https://kubernetes.io/docs/tasks/administer-cluster/reserve-compute-resources/#general-guidelines) suggest a more nuanced approach is recommended?|
+|`ra2.k8s.009`|CPU Pinning|In order to support CPU Pinning, the kubelet must be started with the `--cpu-manager-policy=static` option. (Note, only containers in `Guaranteed` pods - where CPU resource `requests` and `limits` are identical - and configured with positive-integer CPU `requests` will take advantage of this. All other Pods will run on CPUs in the remaining shared pool.)|[infra.com.cfg.003](./chapter02.md#223-cloud-infrastructure-software-profile-requirements)|
```

Comment:
@michaelspedersen - this is the spec I've written for CPU pinning so far.  As mentioned in the other PR, it may be that there are more specific requirements we need to add into chapter02.md in order to drive a more specific decision about using CMK over the native CPU manager.

Reply:
i.e. do we need something in chapter02.md along the lines of "The architecture must support the capability to ensure data plane CNFs are pinned to specific CPU cores"?  To me, the generic "CPU Pinning" requirement from the RM isn't specific enough to help (me anyway) determine whether the native CPU manager is sufficient.

Reply:
in a K8s environment apps should never assume the capability of requesting specific cores
capability to request NUMA aligned cores? sure. and this is supported by K8s CPU Manager
capability to request cores from a special, optimized pool? now this is where it gets tricky :) because DPDK apps might legit need this, but upstream legit does not support this RQ.
ofc you can always separate these apps to their own nodes, and apply low latency tunings to **all** application cores, but that is al ess dynamic approach. and has some OPEX aspect to it (increased consumption, inability to collocate different apps therefore need for more physical servers in the DC etc.)

Reply:
aye - that's what I was starting to think.  Feels like the native CPU manager meets the requirements that are currently defined.

Reply:
I'm wondering what the best approach would be.

The requirement in the RM seems sufficient for VNFs, as the workload only sees the resources that are provided to the VM. i.e. If you assign 3 CPUs (or cores) to the VM, `lscpu` will only show you that many resources. For CNFs the view from inside the container will look like the host, so even with 3 CPUs assigned you might still see 48 CPUs through `lscpu`.

Using [CMK](https://github.com/intel/CPU-Manager-for-Kubernetes) as an example, it adds an extra layer of pinning/isolation through CPU affinity, compared to the native CPU Manager which only limits the processor cycles that each pod has access to. 
My concern is the way that the pinning is enforced. CMK provides a list of assigned cores/threads to the container through environment variables, and these can then be used through either 1) a provided wrapper that sets the affinity of the application or workload running in the container, or 2) to map into an configuration file that accepts a specific cores/threads for processing (i.e. VPP, OVS or generic DPDK applications). This puts responsibility on the workload to abide by these assignments, as I believe it will still be possible to override this assignment.

It might be good to add a requirement to Ch2 that specifies pinning to specific CPUs, but I wonder if it should be a MUST or a SHOULD (especially if it relies on the workload to function properly)

Reply:
> 
> 
> in a K8s environment apps should never assume the capability of requesting specific cores
> capability to request NUMA aligned cores? sure. and this is supported by K8s CPU Manager
> capability to request cores from a special, optimized pool? now this is where it gets tricky :) because DPDK apps might legit need this, but upstream legit does not support this RQ.
> ofc you can always separate these apps to their own nodes, and apply low latency tunings to **all** application cores, but that is al ess dynamic approach. and has some OPEX aspect to it (increased consumption, inability to collocate different apps therefore need for more physical servers in the DC etc.)

I agree that apps shouldn't be able to request specific cores, but as you also mention DPDK apps usually need specific, and typically isolated, cores.

For me the requirement will depend a lot on the target users. For computing this would (almost) never be an issue, but for telcos I can see this being a requirement due to SLAs, in particular ones related to jitter/latency.

Reply:
My understanding (please correct), the native CPU Manager doesn't really offer true isolation among user pods; exclusion is for the Kubernetes processes. CMK, on the other hand, provides isolation even among user pods.

Reply:
> My understanding (please correct), the native CPU Manager doesn't really offer true isolation among user pods; exclusion is for the Kubernetes processes. CMK, on the other hand, provides isolation even among user pods.

Perhaps (I don't know).  Point is, we don't have a requirement stating that is required, we just have one that says CPU pinning is required.  Do we need more detailed requirements such as those I suggest above, or e.g. "The architecture must support the capability for workloads to be scheduled to specific and isolated CPU cores."?

Reply:
@tomkivlin There are additional requirements that do not specify the level of isolation but could still be applicable, for example the following:

Reference | Description | Requirement (common to all Profiles)
-- | -- | --
e.man.004 | Capability to isolate resources between tenants | Must support
sec.wl.006 | Workloads must be separable by their categorisation (for example, payment card information, healthcare, etc.) |

Reply:
IMO you are overmistiyfing this a little bit. as someone who authored one CPU Manager solution, I think the situation is actually simple
if you have any of the following 2 requirements defined:
- possibility for apps to request special exclusive cores, different from other cores (differently tuned than rest on the node)
- possibility for a container to request exclusive **and** shared resources at the same time
than, and only than you need a custom CPU manager.
If you don't have any of these requirements, you don't need a custom CPU manager.

Sidenote: if you need a custom CPU Manager, CMK is not the only option. if you ask me it is not even the best workaround for multiple reasons, but again, I'm biased, obviously (https://github.com/nokia/CPU-Pooler)
at the very minimum just mention that "well, because of these requirements you prob need a special CPU Manager, up to you, just satisfy the requirements"

Reply:
So right now, we don't have either requirement, so the native CPU manager is fine.  I'm cool with that.  My question (which came from some specs that were written in an RI2 PR) was whether the first of your bullets was a requirement, that we hadn't documented yet.  If so, we need to add it to chapter02.md and as you rightly say, deal with it appropriately.  If not, then the RI2 PR needs an update as the CMK isn't required.

Reply:
> My understanding (please correct), the native CPU Manager doesn't really offer true isolation among user pods; exclusion is for the Kubernetes processes. CMK, on the other hand, provides isolation even among user pods.

Upstream CPU manager also can and does isolate exclusive cores from all others.
Only containers running on shared cores, well, share their cores. if your container is sensitive to context switches an argument can be made that you should be requesting exclusive cores in the first place, rather than trying to force the existence of "multiple shared pools" into the infrastructure. If your app is not sensitive to switches, they should be perfectly fine running on a shared cpuset, as there are enough administrative tools in pace to guarantee that you are getting the **amount** of CPU shares you require from the same shared pool.

Reply:
> So right now, we don't have either requirement, so the native CPU manager is fine. I'm cool with that. My question (which came from some specs that were written in an RI2 PR) was whether the first of your bullets was a requirement, that we hadn't documented yet. If so, we need to add it to chapter02.md and as you rightly say, deal with it appropriately. If not, then the RI2 PR needs an update as the CMK isn't required.

I have personally met some data plane applications which did require it. Basically DPDK-only use-case
An another issue if you want satisfy this via providing different exclusive pools on the same node, or optimizing all cores on dedicated nodes as explained above
if it is anyway planned to physically isolate these senstivie workloads then this special pooling might not be required.
on the other hand, there might not be enough nodes isolate these workloads on the far edge
one can argue either way. of course I'm personally well documented and recorded arguing that we need multiple pools, but it certainly has downsides as well (static pool config increase chances of resource fragmentation for example)

Reply:
to put things into real-life context :)
Openshift AFAIK uses upstream CPU Manager only, and solves the issue of data plane apps by 1: defining --reserve-cpus 2: applying the "sepcial config" for all cores (isolcpus basically)
products and projects I'm involved with support the concept of multiple pools, but as an optional, opt-in, per-host group configurable feature (i.e. enable it only on nodes it is required)
not sure about other products, but both approaches are definitely used on the field, so I'm not sure it is wise, or possible to rule out either. flexibility might be the key

Reply:
My takeaway from the discussion is to keep the requirements as already listed (CPU pinning based on native CPU manager). Any additional pinning through e.g. CMK or CPU-Pooler is out of scope, but can be applied as needed. Does this sound right?

Reply:
Yep, I agree with that.